pennylane/gradients/parameter_shift.py

```diff
@@ -327,28 +380,184 @@ def processing_fn(results):
             res = results[start : start + num_tapes] if batch_size is None else results[start]
             start = start + num_tapes
 
-            g = _evaluate_gradient(res, data, broadcast, r0, scalar_qfunc_output)
+            g = _evaluate_gradient(tape, res, data, broadcast, r0, scalar_qfunc_output)
 
             grads.append(g)
             # This clause will be hit at least once (because otherwise all gradients would have
             # been zero), providing a representative for a zero gradient to emulate its type/shape.
             zero_rep = qml.math.zeros_like(g)
 
-        for i, g in enumerate(grads):
-            # Fill in zero-valued gradients
-            if g is None:
-                grads[i] = zero_rep
-            # The following is for backwards compatibility; currently, the device stacks multiple
-            # measurement arrays, even if not the same size, resulting in a ragged array.
-            # In the future, we might want to change this so that only tuples of arrays are returned.
-            if getattr(g, "dtype", None) is np.dtype("object") and qml.math.ndim(g) > 0:
-                grads[i] = qml.math.hstack(g)
+        if qml.active_return():
+            for i, g in enumerate(grads):
+                # Fill in zero-valued gradients
+                if g is None:
+                    grads[i] = zero_rep
 
-        return qml.math.T(qml.math.stack(grads))
+            res = tuple(grads)
+        else:
+            res = _post_process_grads(grads, zero_rep)
+
+        return res
 
     return gradient_tapes, processing_fn
 
 
+def _get_var_with_second_order(pdA2, f0, pdA):
+    """Auxiliary function to compute d(var(A))/dp = d<A^2>/dp -2 * <A> *
+    d<A>/dp for the variances"""
+    return pdA2 - 2 * f0 * pdA
+
+
+def _process_pda2_involutory(tape, pdA2, var_idx):
+    """Auxiliary function for post-processing the partial derivative of <A^2>
+    if there are involutory observables.
+
+    If involutory observables are present, ensure they have zero gradient.
+
+    For the pdA2_tapes, we have replaced non-involutory observables with their
+    square (A -> A^2). However, involutory observables have been left as-is
+    (A), and have not been replaced by their square (A^2 = I). As a result,
+    components of the gradient vector will not be correct. We need to replace
+    the gradient value with 0 (the known, correct gradient for involutory
+    variables).

```

Comment:
I know this was already the case before the return types project, but can you explain why we don't replace the involutory observables in the `pdA2_tapes` by `I`? The gradient will be zero as expected, plus we're computing the expectation values of all the observables in either case, so there's no computational difference.

Reply:
My understanding is that it is a step we can skip because we'd get `0` as a gradient anyways.

One thing I'm not entirely sure about is why we don't simplify tapes to only contain non-involutory observables when we do the 2nd parameter-shift expval computation in `var_param_shift` to compute `pdA2`. Unless I'm missing something, I think that's identical to what's being done here except that we could skip computing incorrect gradients that are replaced by `0` anyways. This is out of scope here though.

Reply:
@antalszava good point, we could do this :)

I think at the time, I must have reasoned that there is very little overhead to those extra (incorrect) gradients since:

- the parameter-shift rule scales wrt input parameters, and is independent of QNode output size
- the gradient computation is vectorized

However, for finite shots/large number of qubits, there likely is overhead for computing expectation values of observables we ultimately don't care about.